src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/bolts/HaFlowDeleteHubBolt.java

```diff
@@ -15,93 +15,190 @@
- */
-public class HaFlowDeleteHubBolt extends AbstractBolt {
-    public static final String COMMON_ERROR_MESSAGE = "Couldn't delete HA-flow";
-    private transient HaFlowRepository haFlowRepository;
-    private transient FlowRepository flowRepository;
+    private transient HaFlowDeleteService service;
+    private String currentKey;
+
+    private LifecycleEvent deferredShutdownEvent;
 
-    public HaFlowDeleteHubBolt(PersistenceManager persistenceManager) {
-        super(persistenceManager);
+    public HaFlowDeleteHubBolt(
+            @NonNull HaFlowDeleteHubBolt.HaFlowDeleteConfig config, @NonNull PersistenceManager persistenceManager,
+            @NonNull FlowResourcesConfig flowResourcesConfig, @NonNull RuleManagerConfig ruleManagerConfig) {
+        super(persistenceManager, config);
+        this.config = config;
+        this.flowResourcesConfig = flowResourcesConfig;
+        this.ruleManagerConfig = ruleManagerConfig;
+        enableMeterRegistry("kilda.ha_flow_delete", HUB_TO_METRICS_BOLT.name());
     }
 
     @Override
     protected void init() {
-        haFlowRepository = persistenceManager.getRepositoryFactory().createHaFlowRepository();
-        flowRepository = persistenceManager.getRepositoryFactory().createFlowRepository();
+        FlowResourcesManager resourcesManager = new FlowResourcesManager(persistenceManager, flowResourcesConfig);
+        RuleManager ruleManager = new RuleManagerImpl(ruleManagerConfig);
+        service = new HaFlowDeleteService(this, persistenceManager, resourcesManager, ruleManager,
+                config.getSpeakerCommandRetriesLimit());
+    }
+
+    @Override
+    protected boolean deactivate(LifecycleEvent event) {
+        if (service.deactivate()) {
+            return true;
+        }
+        deferredShutdownEvent = event;
+        return false;
     }
 
     @Override
-    protected void handleInput(Tuple input) throws PipelineException {
-        String key = pullValue(input, MessageKafkaTranslator.FIELD_ID_KEY, String.class);
-        HaFlowDeleteRequest payload = pullValue(input, FIELD_ID_PAYLOAD, HaFlowDeleteRequest.class);
+    protected void activate() {
+        service.activate();
+    }
 
+    @Override
+    protected void onRequest(Tuple input) throws PipelineException {
+        currentKey = pullKey(input);
+        HaFlowDeleteRequest request = pullValue(input, FIELD_ID_PAYLOAD, HaFlowDeleteRequest.class);
         try {
-            handleHaFlowDelete(key, payload);
-        } catch (FlowProcessingException e) {
-            emitErrorMessage(e.getErrorType(), COMMON_ERROR_MESSAGE, e.getMessage());
-        } catch (Exception e) {
-            emitErrorMessage(ErrorType.INTERNAL_ERROR, COMMON_ERROR_MESSAGE, e.getMessage());
+            service.handleRequest(currentKey, getCommandContext(), request.getHaFlowId());
+        } catch (DuplicateKeyException e) {
+            log.error("Failed to handle a request with key {}. {}", currentKey, e.getMessage());
```

Comment:
I'd add the type into the message itself. Also `key` is countable. 
something like: `Failed to handle a request due to the duplicated request key {}` (or what kind of key is it?)

Reply:
fixed